File: src/cloud/python/ratel_ai_cloud/transport.py
"""Best-effort HTTP transport: batches of events POSTed with retry/backoff."""

from __future__ import annotations

import asyncio
import random
from collections.abc import Awaitable, Callable
from dataclasses import dataclass

import httpx

from .events import Event

#: Upper bound the ingest endpoint accepts in one request.
MAX_BATCH = 500


@dataclass
class SendResult:
    ok: bool
    accepted: int
    status: int | None = None


def _safe_on_error(on_error: Callable[[Exception], None] | None, err: Exception) -> None:
    """Invoke a host error callback without letting it break us: the transport is
    contractually best-effort and must never raise into the host — not even when the
    observer itself throws (which, on the permanent-4xx path, would otherwise be caught
    by the retry loop and turn a permanent drop into repeated retries)."""
    if on_error is None:
        return
    try:
        on_error(err)
    except Exception:
        # A broken observer must not propagate — telemetry is best-effort.
        pass


def _is_retryable(status: int) -> bool:
    return status == 429 or status >= 500


def _backoff(base_delay: float, attempt: int) -> float:
    # Full jitter in the top half of the window keeps a floor while spreading load.
    return float(base_delay * (2**attempt) * (0.5 + random.random() * 0.5))


def _accepted(response: httpx.Response, fallback: int) -> int:
    try:
        data = response.json()
    except ValueError:
        return fallback
    value = data.get("accepted") if isinstance(data, dict) else None
    return value if isinstance(value, int) else fallback


async def send_event_batch(
    events: list[Event],
    *,
    endpoint: str,
    api_key: str,
    client: httpx.AsyncClient | None = None,
    max_retries: int = 3,
    timeout: float = 10.0,
    base_delay: float = 0.2,
    on_error: Callable[[Exception], None] | None = None,
    sleep: Callable[[float], Awaitable[None]] | None = None,
) -> SendResult:
    """POST a batch of events. Best-effort: retries transient failures with
    exponential backoff + jitter, drops on permanent 4xx, and never raises."""
    if not events:
        return SendResult(ok=True, accepted=0)

    retries = max(0, max_retries)
    owns_client = client is None
    # When we own the client the timeout is set at construction; when the caller injects
    # one we respect its configuration rather than overriding per request.
    http = client or httpx.AsyncClient(timeout=timeout)
    nap = sleep or asyncio.sleep
    headers = {"content-type": "application/json", "authorization": f"Bearer {api_key}"}

    try:
        for attempt in range(retries + 1):
            try:
                response = await http.post(endpoint, json=events, headers=headers)
            except httpx.HTTPError as err:
                if attempt == retries:
                    _safe_on_error(on_error, err)
                    return SendResult(ok=False, accepted=0)
                await nap(_backoff(base_delay, attempt))
                continue

            if response.status_code < 300:
                accepted = _accepted(response, len(events))
                return SendResult(ok=True, accepted=accepted, status=response.status_code)

            if not _is_retryable(response.status_code) or attempt == retries:
                _safe_on_error(
                    on_error,
                    RuntimeError(f"ratel-cloud: ingest rejected with {response.status_code}"),
                )
                return SendResult(ok=False, accepted=0, status=response.status_code)

            await nap(_backoff(base_delay, attempt))

        return SendResult(ok=False, accepted=0)
    finally:
        if owns_client:
            await http.aclose()

File: src/cloud/ts/src/transport.ts
import type { Event } from "./types.js";

/** Upper bound the ingest endpoint accepts in one request. */
export const MAX_BATCH = 500;

export interface TransportOptions {
  endpoint: string;
  apiKey: string;
  /** Injectable for testing; defaults to the global `fetch`. */
  fetch?: typeof fetch;
  /** Retries after the first attempt for transient failures. Default 3. */
  maxRetries?: number;
  /** Per-request timeout in ms. Default 10_000. */
  timeoutMs?: number;
  /** Base backoff delay in ms (doubled per attempt, jittered). Default 200. */
  baseDelayMs?: number;
  /** Called with any swallowed error so the host can observe failures. */
  onError?: (err: unknown) => void;
  /** Injectable sleep, for testing. */
  sleep?: (ms: number) => Promise<void>;
}

export interface SendResult {
  ok: boolean;
  accepted: number;
  status?: number;
}

const defaultSleep = (ms: number) => new Promise<void>((resolve) => setTimeout(resolve, ms));

/** Transient HTTP statuses worth retrying. */
function isRetryableStatus(status: number): boolean {
  return status === 429 || status >= 500;
}

/**
 * Invoke a host-supplied error callback without ever letting it break us. The client
 * is contractually "never throws into the host"; a callback that throws must not defeat
 * that (nor, on the 4xx path, be caught by the retry loop and turn a permanent drop into
 * repeated retries).
 */
export function safeOnError(onError: ((err: unknown) => void) | undefined, err: unknown): void {
  if (!onError) return;
  try {
    onError(err);
  } catch {
    // A broken observer must not propagate — telemetry is best-effort.
  }
}

/**
 * POST a batch of events to the ingest endpoint. Best-effort: retries transient
 * failures with exponential backoff + jitter (honoring `Retry-After` on 429), drops
 * on permanent 4xx, and **never throws** — failures are reported via `onError` and a
 * falsy result.
 */
export async function sendEventBatch(events: Event[], opts: TransportOptions): Promise<SendResult> {
  if (events.length === 0) return { ok: true, accepted: 0 };

  const doFetch = opts.fetch ?? fetch;
  const maxRetries = Math.max(0, opts.maxRetries ?? 3);
  const timeoutMs = opts.timeoutMs ?? 10_000;
  const baseDelayMs = opts.baseDelayMs ?? 200;
  const sleep = opts.sleep ?? defaultSleep;
  const body = JSON.stringify(events);

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    let retryAfterMs: number | undefined;
    try {
      // A fresh AbortController per attempt, cleared as soon as the fetch settles, so no
      // timer lingers after a fast response (unlike `AbortSignal.timeout`).
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), timeoutMs);
      let response: Response;
      try {
        response = await doFetch(opts.endpoint, {
          method: "POST",
          headers: {
            "content-type": "application/json",
            authorization: `Bearer ${opts.apiKey}`,
          },
          body,
          signal: controller.signal,
        });
      } finally {
        clearTimeout(timer);
      }

      if (response.ok) {
        const accepted = await readAccepted(response, events.length);
        return { ok: true, accepted, status: response.status };
      }

      if (!isRetryableStatus(response.status) || attempt === maxRetries) {
        safeOnError(
          opts.onError,
          new Error(`ratel-cloud: ingest rejected with ${response.status}`),
        );
        return { ok: false, accepted: 0, status: response.status };
      }
      retryAfterMs = parseRetryAfter(response.headers.get("retry-after"));
    } catch (err) {
      if (attempt === maxRetries) {
        safeOnError(opts.onError, err);
        return { ok: false, accepted: 0 };
      }
    }

    await sleep(retryAfterMs ?? backoffDelay(baseDelayMs, attempt));
  }

  return { ok: false, accepted: 0 };
}

async function readAccepted(response: Response, fallback: number): Promise<number> {
  try {
    const data = (await response.json()) as { accepted?: number };
    return typeof data.accepted === "number" ? data.accepted : fallback;
  } catch {
    return fallback;
  }
}

/** `Retry-After` in delta-seconds → ms; HTTP-date and malformed values fall back to backoff. */
function parseRetryAfter(header: string | null): number | undefined {
  if (!header) return undefined;
  const seconds = Number(header);
  return Number.isFinite(seconds) && seconds >= 0 ? seconds * 1000 : undefined;
}

function backoffDelay(baseDelayMs: number, attempt: number): number {
  const exponential = baseDelayMs * 2 ** attempt;
  // Full jitter in the top half of the window keeps a floor while spreading load.
  return exponential * (0.5 + Math.random() * 0.5);
}
